thavalon-server/src/game/engine.rs

```diff
@@ -0,0 +1,761 @@
+            if agravaine_declared {
+                log::debug!("Agravaine declared!");
+
+                // maybe this should be a different message?
+                self.interactions
+                    .send(Message::MissionResults {
+                        passed: false,
+                        successes: cards.values().filter(|c| c == &&Card::Success).count(),
+                        fails: cards.values().filter(|c| c == &&Card::Fail).count(),
+                        reverses: cards.values().filter(|c| c == &&Card::Reverse).count(),
+                    })
+                    .await?;
+
+                return Ok(false);
+            }
+        }
+
+        Ok(!failed)
+    }
+}
+
+// Maybe a typemap for per-role game state?
+// can have a base trait for callbacks (ex. reset Maeve's counter every round)
+
+/// Tests if a mission has failed
+fn is_failure<'a, I: IntoIterator<Item = &'a Card>>(
+    spec: &GameSpec,
+    mission: usize,
+    cards: I,
+) -> bool {
+    let (fails, reverses) = cards
+        .into_iter()
+        .fold((0, 0), |(fails, reverses), card| match card {
+            Card::Fail => (fails + 1, reverses),
+            Card::Reverse => (fails, reverses + 1),
+            Card::Success => (fails, reverses),
+        });
+
+    // Two reverses cancel each other out
+    let reversed = reverses % 2 == 1;
+
+    if mission == 4 && spec.double_fail_mission_four {
+        // Mission 4 can fail if there are 2+ fails or a reverse and a fail, in large enough games
+        (fails >= 2 && !reversed) || (fails == 1 && reverses == 1)
+    } else {
+        // Normally, missions fail if they contain a fail OR are reversed
+        (fails > 0) ^ reversed
+    }
+}
+
+impl VotingResults {
+    fn new(votes: HashMap<PlayerId, bool>) -> VotingResults {
+        let mut net = 0; // Will be above 0 if there are more upvotes than downvotes
+        for vote in votes.values() {
+            if *vote {
+                net += 1;
+            } else {
+                net -= 1;
+            }
+        }
+
+        VotingResults {
+            votes,
+            sent: net > 0,
+        }
+    }
+
+    /// Players who upvoted the proposal
+    fn upvotes<'a>(&'a self) -> impl Iterator<Item = PlayerId> + 'a {
+        self.votes
+            .iter()
+            .filter_map(|(player, vote)| if *vote { Some(*player) } else { None })
+    }
+
+    /// Players who downvoted the proposal
+    fn downvotes<'a>(&'a self) -> impl Iterator<Item = PlayerId> + 'a {
+        self.votes
+            .iter()
+            .filter_map(|(player, vote)| if !*vote { Some(*player) } else { None })
+    }
+}
+
+impl<'a> fmt::Display for Proposal<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.players
+            .iter()
+            .format_with(", ", |id, f| f(&self.game.name(*id)))
+            .fmt(f)?;
+        write!(f, " (proposed by {})", self.game.name(self.proposer))
+    }
+}
+
+#[test]
+fn test_is_failure() {
```

Comment:
Is there a reason these are outside of the `tests{}` namespace?

Reply:
Because I added the `test` namespace after writing this test :upside_down_face: